src/filters/sse.rs

```diff
@@ -52,119 +53,121 @@ use futures::{future, Stream, TryStream, TryStreamExt};
 use http::header::{HeaderValue, CACHE_CONTROL, CONTENT_TYPE};
 use hyper::Body;
 use pin_project::pin_project;
-use serde::Serialize;
-use serde_json;
+use serde_json::{self, Error};
 use tokio::time::{self, Delay};
 
-use self::sealed::{
-    BoxedServerSentEvent, EitherServerSentEvent, SseError, SseField, SseFormat, SseWrapper,
-};
+use self::sealed::SseError;
 use super::header;
 use crate::filter::One;
 use crate::reply::Response;
 use crate::{Filter, Rejection, Reply};
 
-/// Server-sent event message
-pub trait ServerSentEvent: SseFormat + Sized + Send + Sync + 'static {
-    /// Convert to either A
-    fn into_a<B>(self) -> EitherServerSentEvent<Self, B> {
-        EitherServerSentEvent::A(self)
+// Server-sent event data type
+#[derive(Debug)]
+enum DataType {
+    Text(String),
+    Json(String),
+}
+
+/// Server-sent event
+#[derive(Default, Debug)]
+pub struct Event {
+    name: Option<String>,
+    id: Option<String>,
+    data: Option<DataType>,
+    event: Option<String>,
+    comment: Option<String>,
+    retry: Option<Duration>,
+}
+
+impl Event {
+    /// Set Server-sent event data
+    /// data field(s) ("data:<content>")
+    pub fn data<T: Display>(mut self, data: Option<T>) -> Event {
+        self.data = data.map(|d| DataType::Text(d.to_string()));
+        self
     }
 
-    /// Convert to either B
-    fn into_b<A>(self) -> EitherServerSentEvent<A, Self> {
-        EitherServerSentEvent::B(self)
+    /// Set Server-sent event data
+    /// data field(s) ("data:<content>")
+    pub fn json_data<T: Serialize>(mut self, data: Option<T>) -> Result<Event, Error> {
+        self.data = match data {
+            Some(data) => Some(DataType::Json(serde_json::to_string(&data)?)),
+            None => None,
+        };
+        Ok(self)
     }
 
-    /// Convert to boxed
-    fn boxed(self) -> BoxedServerSentEvent {
-        BoxedServerSentEvent(Box::new(self))
+    /// Set Server-sent event comment
+    /// Comment field (":<comment-text>")
+    pub fn comment<T: Display>(mut self, comment: Option<T>) -> Event {
+        self.comment = comment.map(|c| c.to_string());
+        self
     }
-}
 
-impl<T: SseFormat + Send + Sync + 'static> ServerSentEvent for T {}
+    /// Set Server-sent event event
+    /// Event name field ("event:<event-name>")
+    pub fn event<T: Display>(mut self, event: Option<T>) -> Event {
+        self.event = event.map(|e| e.to_string());
+        self
+    }
 
-#[allow(missing_debug_implementations)]
-struct SseComment<T>(T);
+    /// Set Server-sent event retry
+    /// Retry timeout field ("retry:<timeout>")
+    pub fn retry(mut self, duration: Option<Duration>) -> Event {
```

Comment:
Is there a strong reason to keep these arguments as `Option`? With a builder, will people likely (ever) need to set a property, and later on the same builder, unset it?

Reply:
following your question on https://github.com/seanmonstar/warp/issues/645#issue-655253240 I have the idea we spoke about making them all `Option` but I am not sure now aha :sweat_smile: , should we define mandatory fields, and/or implement  mutual exclusivity as refered on https://github.com/seanmonstar/warp/pull/663#discussion_r541655071?

Reply:
Well, they probably should be an option internally. My comment just now was that the builder function arguments probably don't need to take an `Option`.

Reply:
ah right! yeah I think I the reason was allowing to unset, but i doubt people will need to set a property and later unset it.
Updated it!